cvpaperchallenge/Crux
# #14/chore/implement ci

## Issue URL

#14 

## Change overview

- Replace`docker-compose` with `docker compose` as `docker-compose` will be unsupported at the end of June 2023 [[link](https://docs.docker.com/compose/install/)] 
   - `ubuntu-latest` image comes standard with `docker compose` [[link](https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2204-Readme.md)]
- Remove the `--parallel` option from the `docker compose build` command as it is set in the `docker compose` command level (the official Docs says `--parallel` defaults to `-1`(unlimited) [[link](https://docs.docker.com/engine/reference/commandline/compose/#options)])
- Change the `set-output` notation to use  `GITHUB_OUTPUT` as `set-output` will cause an error after June 2023 [[link](https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/)] (but it's not accessed in subsequent processes, so may as well take it away)
- Change to get `UID` and `GID` from the host image, since those numbers of GitHub Actions images are sometimes not 1000
- Update the cache process for the Crux project specs to make it easier to work on it later

## How to test

1. Fork this project
2. Raise the pull request from `#14/chore/implement-ci` to `develop` on the forked repository
3. Merge the pull request
4. Make a small change on the `develop` branch and commit it to the new branch
5. Raise the pull request from that new branch to `develop`
6. Open the Actions tab and check the CI works without any errors

## Note for reviewers

I implemented the very basic cache features, but some problem still remains.

1. When the trigger is `pull_request`, a cache is attached to `refs/pull/{pull_request_number}/merge`, so the cache is valid for operations **only on that pull request**, not on the base branch or any other pull requests targeting that base branch [[link](https://docs.github.com/ja/actions/using-workflows/caching-dependencies-to-speed-up-workflows#restrictions-for-accessing-a-cache)]
2. Past caches are not deleted (the CI process triggered by the updates to the same pull request will successfully use the past cache, but a new cache will be created and do not overwrite the old one)
  - Running log for the 1st time
    ![image](https://user-images.githubusercontent.com/61688218/236384167-b44f4efe-0dfc-4a06-947a-7caadf16979c.png)
  - Running log for the 2nd time (successfully receive cache)
    ![image](https://user-images.githubusercontent.com/61688218/236383980-01f7f8b0-54f0-4681-8a9a-b3aa41287b9c.png)
  - Duplicated cache
    ![image](https://user-images.githubusercontent.com/61688218/236383811-5bb5173c-e4c7-42cb-8029-6cf8f585ee97.png)
